### ISSUE TITLE
minor cleanups/fixes

### ISSUE THREAD

**Issue by afilinski** [Ginko-X/Streaming_NESL#17]
* modify rin to have type Int -> Proc AVal, signal error on premature EOS
* remove dependency on tree2Sids in interpreter by changing type of SFun?
* signal runtime error instead of divergence on deadlock
* general code cleanup/simplification/commenting/documentation...
    * SvcodeProcInterp getting somewhat hard to follow

**Comment by Ginko-X** [comment]
- done
- In streaming interpretation, we do only need the SIds extracted from the `stree` inside both `SFun _ stree _ _` and `WithCtrl _  _ stree`, but we need the structure/type information carried by `stree` to construct the final output, i.e, transform `[[AVals]]` to **some** `SvVal`. Also, in the fully eager interpreter, `stree` tells us which type of empty streams we should generate, `SBVal []` or `SIVal []`. So there should be some more changes if we want to remove the dependency.
 - added a deadlock detecting function (but can not guarantee 100% correctness as Procs are not comparable) 
- see #21

**Comment by Ginko-X** [comment]
As I remember, last week we also discussed this bullet 2 again, i.e., 
```Haskell
-- change from
data SFun = SFun [SId] STree [SInstr] Int -- current SFun
-- to
data SFun = SFun [SId] [SId] [SInstr] Int  -- new SFun
```
For WithCtrl, It should be feasible to change its STree to [SId] if we add some `PType` to SVCODE instructions when necessary.  
But for SFun, 
> we need the structure/type information carried by stree to construct the final output

For example, if we know the high-level type is a pair (\tau_1, \tau_2), and we get a list of streams from running SVCODE (by looking up these SIds in the final store), then we still don't know how to transback these streams correctly to some high-level value. 
I think we should at least maintain the binary tree information (instead of SSTr and SPTr) in SFun, not just a list of SIds.

**Comment by afilinski** [comment]
Yes, you'll almost certainly still need to maintain this information somewhere, but it should be outside the modules specifying the syntax and semantics of Svcode itself. The Svcode interpreter (whether eager or streaming) doesn't need the information about how the primitive streams are combined to represent a Snesl sequence with multiple layers of flags; it just evaluates all the streams and hands them off to the printer, whose job it is to translate them  back to to Snesl values. (This is very much like in the formalization of the toy language, where the low-level register code had no knowledge of, or special support for, tuples in the source language.)

I think the source of the confusion may be that you are currently using a dummy SFun in runCompileExp to hold the code for the top-level expression to be evaluated, even though that has nothing to do with function definitions.  If you introduce a new datatype for use with runCompileExp, you should be able to reserve SFun to only be used in FEnv, and make it match the SCall instruction, i.e., with both intputs and outputs being flat lists of SIds.

**Comment by Ginko-X** [comment]
Ok, makes sense. Thank you!
